libraries/standard/mqtt/cbmc/sources/mqtt_cbmc_state.c

```diff
@@ -18,13 +18,139 @@
+{
+    if( pPacketInfo == NULL )
+    {
+        pPacketInfo = mallocCanFail( sizeof( MQTTPacketInfo_t ) );
+    }
+
+    if( pPacketInfo != NULL )
+    {
+        __CPROVER_assume( pPacketInfo->remainingLength < CBMC_MAX_OBJECT_SIZE );
+        pPacketInfo->pRemainingData = mallocCanFail( pPacketInfo->remainingLength );
+    }
+
+    return pPacketInfo;
+}
+
+bool isValidMqttPacketInfo( const MQTTPacketInfo_t * pPacketInfo )
+{
+    bool isValid = true;
+
+    if( pPacketInfo != NULL )
+    {
+        isValid = pPacketInfo->remainingLength < CBMC_MAX_OBJECT_SIZE;
+    }
+
+    return isValid;
+}
+
+MQTTPublishInfo_t * allocateMqttPublishInfo( MQTTPublishInfo_t * pPublishInfo )
+{
+    if( pPublishInfo == NULL )
+    {
+        pPublishInfo = mallocCanFail( sizeof( MQTTPublishInfo_t ) );
+    }
+
+    if( pPublishInfo != NULL )
+    {
+        __CPROVER_assume( pPublishInfo->topicNameLength < CBMC_MAX_OBJECT_SIZE );
+        pPublishInfo->pTopicName = mallocCanFail( pPublishInfo->topicNameLength );
+        __CPROVER_assume( pPublishInfo->payloadLength < CBMC_MAX_OBJECT_SIZE );
+        pPublishInfo->pPayload = mallocCanFail( pPublishInfo->payloadLength );
+    }
+
+    return pPublishInfo;
+}
+
+bool isValidMqttPublishInfo( const MQTTPublishInfo_t * pPublishInfo )
+{
+    bool isValid = true;
+
+    if( pPublishInfo != NULL )
+    {
+        bool validQos = ( ( pPublishInfo->qos >= MQTTQoS0 ) ||
+                          ( pPublishInfo->qos <= MQTTQoS2 ) );
+
+        bool validTopicNameLength = pPublishInfo->topicNameLength < CBMC_MAX_OBJECT_SIZE;
+        bool validPayloadLength = pPublishInfo->payloadLength < CBMC_MAX_OBJECT_SIZE;
+
+        isValid = validQos && validTopicNameLength && validPayloadLength;
+    }
+
+    return isValid;
+}
+
+MQTTConnectInfo_t * allocateMqttConnectInfo( MQTTConnectInfo_t * pConnectInfo )
+{
+    if( pConnectInfo == NULL )
+    {
+        pConnectInfo = mallocCanFail( sizeof( MQTTConnectInfo_t ) );
+    }
+
+    if( pConnectInfo != NULL )
+    {
+        __CPROVER_assume( pConnectInfo->clientIdentifierLength < CBMC_MAX_OBJECT_SIZE );
+        pConnectInfo->pClientIdentifier = mallocCanFail( pConnectInfo->clientIdentifierLength );
+        __CPROVER_assume( pConnectInfo->userNameLength < CBMC_MAX_OBJECT_SIZE );
+        pConnectInfo->pUserName = mallocCanFail( pConnectInfo->userNameLength );
+        __CPROVER_assume( pConnectInfo->passwordLength < CBMC_MAX_OBJECT_SIZE );
+        pConnectInfo->pPassword = mallocCanFail( pConnectInfo->passwordLength );
+    }
+
+    return pConnectInfo;
+}
+
+bool isValidMqttConnectInfo( const MQTTConnectInfo_t * pConnectInfo )
+{
+    bool isValid = true;
+
+    if( pConnectInfo != NULL )
+    {
+        isValid = isValid && ( pConnectInfo->clientIdentifierLength < CBMC_MAX_OBJECT_SIZE );
+        isValid = isValid && ( pConnectInfo->userNameLength < CBMC_MAX_OBJECT_SIZE );
+        isValid = isValid && ( pConnectInfo->passwordLength < CBMC_MAX_OBJECT_SIZE );
```

Comment:
Does `isValid &= ( pConnectInfo->passwordLength < CBMC_MAX_OBJECT_SIZE );` also work?

Reply:
No that is a bitwise and